aten/src/ATen/native/transformers/cuda/sdp_utils.h

```diff
@@ -0,0 +1,241 @@
+    return false;
+  }
+  return true;
+}
+
+inline bool check_for_attn_mask(sdp_params params, bool debug) {
+  if (params.has_attn_mask) {
+    TORCH_CHECK(debug, "Flash Attention does not support attention mask.");
+    return false;
+  }
+  return true;
+}
+
+inline bool check_tensor_shapes(sdp_params params, bool debug) {
+  if (params.query.dim() != 4) {
+    TORCH_CHECK(
+        debug,
+        "Query is not a 4 dimensional tensor. Query dim: ",
+        params.query.dim());
+    return false;
+  }
+  if (params.key.dim() != 4) {
+    TORCH_CHECK(
+        debug,
+        "Key is not a 4 dimensional tensor. Key dim: ",
+        params.key.dim());
+    return false;
+  }
+  if (params.value.dim() != 4) {
+    TORCH_CHECK(
+        debug,
+        "Value is not a 4 dimensional tensor. Value dim: ",
+        params.value.dim());
+    return false;
+  }
+  return true;
+}
+
+inline bool check_head_dim_size(
+    sdp_params params,
+    const std::unordered_set<int64_t> allowed_sizes,
+    bool debug) {
+  if (allowed_sizes.find(params.query.size(-1)) == allowed_sizes.end()) {
+    TORCH_CHECK(
+        debug,
+        "Query's last dimension is not in the set of supported sizes. Query.size(-1): ",
+        params.query.size(-1));
+    return false;
+  }
+  if (allowed_sizes.find(params.key.size(-1)) == allowed_sizes.end()) {
+    TORCH_CHECK(
+        debug,
+        "Key's last dimension is not in the set of supported sizes. Key.size(-1): ",
+        params.key.size(-1));
+    return false;
+  }
+  if (allowed_sizes.find(params.value.size(-1)) == allowed_sizes.end()) {
+    TORCH_CHECK(
+        debug,
+        "Value's last dimension is not in the set of supported sizes. Value.size(-1): ",
+        params.value.size(-1));
+    return false;
+  }
+  return true;
+}
+
+inline bool check_runtime_disabled(sdp_params params, bool debug) {
+  // We check the global context to see if user has explicitly turned of flash sdp kernels
+  auto& ctx = at::globalContext();
+  if (!ctx.userEnabledFlashSDP()) {
+    TORCH_CHECK(debug, "Flash attention has been runtime disabled.");
+    return false;
+  }
+  return true;
+}
+
+inline bool check_gpu_sm75_or_greater(sdp_params params, bool debug) {
+  // Check that the gpu is capable of running flash attention
+  auto dprops = at::cuda::getCurrentDeviceProperties();
+  bool is_sm75 = dprops->major == 7 && dprops->minor == 5;
+  bool is_sm8x = dprops->major == 8 && dprops->minor >= 0;
+  if (!(is_sm8x || is_sm75)) {
+    TORCH_CHECK(
+        debug,
+        "Flash attention only supports sm75 and sm8x gpu architectures. Attempting to run on a sm ",
+        dprops->major,
+        ".",
+        dprops->minor,
+        " gpu.");
+    return false;
+  }
+  return true;
+}
+
+inline bool use_flash_attention(sdp_params params, bool debug) {
+  // Constraints specific to flash attention
+  static const std::unordered_set<int64_t> flash_embed_dim_sizes{
+      16, 32, 64, 128};
```

Comment:
I think with the refactor all head_dims%8 ==0 are supported but let me check and update if thats true